### DISCUSSION TITLE
How to restore a soft deleted record?

### DISCUSSION THREAD

**Discussion by SorenBech2** [Azure/azure-mobile-apps#782] [Q&A] [answered]
I need to be able to restore/undelete a record which has been deleted using softdelete.

The record is deleted on the client side and the table is synchronized with the server (tableName.PushItemsAsync()). The "deleted" field on the server is correctly set to 1 and the record is removed from the offline table (as far as I can tell).

I have been unable to find a undelete function/attribute in the Azure Mobile Apps documentation, so my working hypothesis is, that the way to do this is to change the Deleted value from 1 to 0, by pulling the record from the server and change the Deleted value in the offline table and then sync with the server.

Question 1: Is this the correct way of undeleting a record that has been soft-deleted?

If so, this requires a pull of the soft deleted record to the offline table - and I have not been able to make this work.

Question 2: How do I pull a soft-deleted record to the offline table?

I have tried the following:
Prior to deleting the record in the offline table I save it to a variable "record".
I then try to pull the record from the server using:

var queryName = tableName.CreateQuery().Where(x => x.Id == record.Id).IncludeDeletedItems();
await tableName.PullItemsAsync(queryName);

I have tried also tried with the below variations:

var queryName = tableName.CreateQuery().Where(x => x.Id == record.Id).IncludeDeletedItems(true);
await tableName.PullItemsAsync(queryName);

var queryName = tableName.CreateQuery().Where(x => x.Deleted == true);
await tableName.PullItemsAsync(queryName);

await tableName.PullItemAsync(query: $"$filter=Deleted eq true");

and a few other variations but none of them the pulls the soft deleted record(s) to the offline table. 

I'm running out of good ideas, so I really hope you can help.

Thanks in advance.

**Comment by adrianhall** [comment, marked as answer]
Short version is that it cannot be done via the offline table.  You must do it via the REST API. 

Use the following snippet (given a remoteTable which is a IRemoteTable<T>):

```csharp
var item = await remoteTable.GetItemAsync(id);
await remoteTable.UndeleteItemAsync(item);
```

Enjoy!

**Reply by SorenBech2** [reply]
Hi Adrian

Thanks for rapid reply. Very appriciated.

I can't seem to get it to work though. Using your snippet 

                var item = await remoteTable.GetItemAsync(Id);
                await remoteTable.UndeleteItemAsync(item);

I get the following results:

var item = await remoteTable.GetItemAsync(Id); (where Id is the Id of the record in the remote table marked deleted) 
yields:
DatasyncInvalidOperationException: The request could not be completed (Gone)

If I save the entire record in "item" prior to deleting it and use this record in the 
  await remoteTable.UndeleteItemAsync(item);
I get:
DatasyncConflictException: The request could not be completed (pre-condition failed)
(whish I expected since I assume the Version and UpdatedAt will be out of sync - but had to try).

the remoteTable handle works fine on non-deleted items on the server (count, get).

(It is as if I'm missing a "IncludeDeletedItems" somewhere, but I cannot figure out where.) 

Can you help pls.?

**Reply by adrianhall** [reply]
I'm going to turn this into an issue for a feature request.

Ideally, `remoteTable.GetItemAsync(id, includeDeletedItems: true)` is what we want, but that doesn't exist.

**Reply by adrianhall** [reply]
In the interim, try this:

```csharp
var item = (await remoteTable.Where(x => x.Id == $id).IncludeDeletedItems().ToListAsync()).SingleOrDefault();
```

The `.IncludeDeletedItems()` does the request to, well, include the soft-deleted items.  You can also do something like:

```csharp
var allDeletedItems = await remoteTable.Where(x => x.Deleted).IncludeDeletedItems().ToListAsync();
```

Let me know how it goes.

**Reply by SorenBech2** [reply]
Exactly what is needed. Would be great to have that feature.
Your work around suggestions works like a charm. Thanks a bunch for your fast and good help.

**Reply by adrianhall** [reply]
Fixed in 6.0.4, which I just cut for release.

**Reply by SorenBech2** [reply]
Installed and working. Thank you Adrian. Very impressive service.
